File: frontend/src/api/policy.ts
// The autonomy-tier API (issue #562): the console reads and writes the
// company's effective `[policy]` through the host's `.../policy` routes (REST,
// camelCase over the wire).
//
// The effective policy is the operator's console override where it sets a
// field, and the committed manifest `[policy]` everywhere else. The console
// never writes the manifest — a rebuild re-persists that from the seed, and for
// `[policy]` that is a deliberate security property — so a change here is a
// durable, attributed *override* the host resolves ahead of the manifest.
//
// Standalone functions over the shared client (mirrors `api/inference.ts`), so
// no change to `OpenCompanyClient` or the shared `api/types.ts` is needed.

import type { OpenCompanyClient } from "./client";

/**
 * One selectable tier, with the host's own words for what it means.
 *
 * The prose comes from the host rather than living here on purpose: it
 * describes what this runtime's approval gate actually does, and a copy in
 * TypeScript would drift from the behaviour it claims to describe. The list
 * also only contains tiers the host accepts, so a console built against a newer
 * or older host offers exactly what that host can honour.
 */
export interface PolicyTier {
  /** The `[policy].mode` word. */
  value: string;
  /** The operator-facing label. */
  label: string;
  /** What choosing it means, in consequences rather than tier vocabulary. */
  description: string;
}

/** The company's effective policy, plus what a reset would restore. */
export interface PolicyStatus {
  /** The tier actually in force. */
  mode: string;
  /**
   * The always-ask list actually in force — the operator's real lever. It wins
   * over every tier, `full` included.
   */
  alwaysApprove: string[];
  /** Spend strictly under this amount without an approval; `null` means no cap. */
  autoApproveUnderUsd: number | null;
  /** How long an undecided approval remains actionable. */
  approvalTtlHours: number;
  /** The manifest's tier, so "reset" can name what it would restore. */
  manifestMode: string;
  /** The manifest's always-ask list, for the same reason. */
  manifestAlwaysApprove: string[];
  /** The manifest's spend cap before a console override. */
  manifestAutoApproveUnderUsd: number | null;
  /** The manifest's configured deadline, or `null` when it uses the default. */
  manifestApprovalTtlHours: number | null;
  /**
   * Whether an operator override is in force.
   *
   * Deliberately not derivable by comparing the values: an override that
   * happens to match the manifest is still an override, and is still what a
   * reset would remove.
   */
  overridden: boolean;
  /** Who set the override, if one is set. */
  setBy?: string;
  /** When it was set (epoch millis), if one is set. */
  setAtMillis?: number;
  /** The selectable tiers, in increasing order of autonomy. */
  tiers: PolicyTier[];
  /**
   * When a change bites, in the host's words.
   *
   * Rendered rather than paraphrased: a tier change lands on the company's
   * NEXT turn, so a turn already running finishes under the previous tier.
   * Since "stop the flood now" is what an operator comes here to do, that gap
   * is worth stating rather than leaving them to discover.
   */
  takesEffect: string;
  /**
   * Every tool name this build's approval gate can match — the complete
   * registry, not the workflow-authorable subset served by
   * `/workflows/tool-slugs`. The "is this a real tool?" note under the field
   * compares against this when the host serves it, so a wired agent tool
   * (`hosting_launch_site`, `publish_artifact`) is never called a mistake just
   * because it cannot be a workflow node. Absent on a host predating the field.
   */
  knownTools?: string[];
}

/**
 * The set-policy body. Omit a field to leave it alone; send `null` to stop
 * overriding it.
 *
 * `alwaysApprove: []` is an operator deliberately clearing the always-ask list,
 * which is NOT the same as omitting the field. Sending neither field is a 422
 * rather than a silent no-op.
 */
export interface SetPolicyInput {
  mode?: string | null;
  alwaysApprove?: string[] | null;
  /** `null` means no spend cap; omit to leave the cap alone. */
  autoApproveUnderUsd?: number | null;
  /** `null` stops overriding the deadline; omit to leave it alone. */
  approvalTtlHours?: number | null;
}

/** The company's effective policy. */
export function getPolicy(
  client: OpenCompanyClient,
  company: string | null,
): Promise<PolicyStatus> {
  return client.get<PolicyStatus>(`${client.scopeFor(company)}/policy`);
}

/** Set the tier and/or the always-ask list. Admin-only, attributed. */
export function setPolicy(
  client: OpenCompanyClient,
  company: string | null,
  body: SetPolicyInput,
): Promise<PolicyStatus> {
  return client.put<PolicyStatus>(`${client.scopeFor(company)}/policy`, body);
}

/** Drop the override so the manifest's `[policy]` applies again. */
export function resetPolicy(
  client: OpenCompanyClient,
  company: string | null,
): Promise<PolicyStatus> {
  return client.del<PolicyStatus>(`${client.scopeFor(company)}/policy`);
}

File: src/server/graphql/policy.rs
//! The autonomy tier over GraphQL (issue #1070) — read-only.
//!
//! # Why this exists at all
//!
//! `Company` already resolves `approvals` and `pendingApprovals`, and did not
//! resolve the setting that decides whether anything ever *enters* that queue.
//! An empty `approvals` is either a healthy `supervised` company with nothing
//! pending or a company on `full` where nothing will ever park, and over
//! GraphQL alone those were indistinguishable.
//!
//! # Why it answers from the REST derivation rather than the record
//!
//! Every value here comes from
//! [`PolicyDto::build`](crate::server::ops::policy::PolicyDto::build), the same
//! function `GET {scope}/policy` answers with, which in turn reads
//! [`CompanyRecord::effective_policy`] — the resolver the approval gate itself
//! uses (`workflows::caps`, `harness`).
//!
//! Recomputing the tier here would have been three lines and a defect. A
//! company carrying a console override would report the overridden tier on one
//! surface and the manifest's on the other, with no way for a caller to know
//! which it received — the two-sources-of-truth failure issue #1027 nearly
//! shipped by adding this value to `/capabilities` instead. The `approvals`
//! field's own header states the same rule for redaction (issue #618): a
//! treatment applied to the REST handler alone leaves the boundary reachable
//! through one GraphQL field.
//!
//! # Read-only, deliberately
//!
//! There is no mutation. Changing the tier through an API is a separate and
//! much larger question, and #1027 disclaimed it for REST on the same grounds.
//! The write plane that does exist (`PUT`/`DELETE {scope}/policy`, issue #562)
//! is admin-gated and attributed; a GraphQL mutation would have to reproduce
//! both, which is not a thing to do as a side effect of closing a read gap.

use std::sync::Arc;

use async_graphql::SimpleObject;

use crate::runtime::CompanyRuntime;
use crate::server::ops::policy::PolicyDto;

/// One selectable tier and what choosing it means.
#[derive(Debug, SimpleObject)]
#[graphql(name = "PolicyTier")]
pub struct PolicyTierGql {
    /// The `[policy].mode` word.
    pub value: String,
    /// The operator-facing label.
    pub label: String,
    /// What choosing it means, in consequences rather than tier vocabulary.
    pub description: String,
}

/// The company's approval policy, as `GET {scope}/policy` reports it.
#[derive(Debug, SimpleObject)]
#[graphql(name = "Policy")]
pub struct PolicyGql {
    /// The tier **actually in force** — the console override where one sets it,
    /// the committed manifest otherwise.
    pub mode: String,
    /// The always-ask list actually in force. The operator's real lever: it
    /// wins over every tier, `full` included.
    pub always_approve: Vec<String>,
    /// The spend cap actually in force — console override where one sets it,
    /// the committed manifest otherwise. A higher cap is looser; `null` is the
    /// strictest setting — no spend is auto-approved, so every spend parks.
    pub auto_approve_under_usd: Option<f64>,
    /// How long an undecided approval remains actionable, in hours.
    pub approval_ttl_hours: f64,
    /// The manifest's tier, so a client can see what a reset would restore.
    pub manifest_mode: String,
    /// The manifest's always-ask list, for the same reason.
    pub manifest_always_approve: Vec<String>,
    /// The manifest's spend cap, for the same reason — what a reset would
    /// restore (a higher cap is looser; `null` is the strictest setting — no
    /// spend is auto-approved, so every spend parks).
    pub manifest_auto_approve_under_usd: Option<f64>,
    /// The manifest's approval deadline, for the same reason — what a reset
    /// would restore.
    pub manifest_approval_ttl_hours: Option<f64>,
    /// Whether an operator override is in force.
    ///
    /// Distinct from comparing `mode` with `manifestMode`: an override that
    /// happens to match the manifest is still an override, and is still what a
    /// reset would remove.
    pub overridden: bool,
    /// Who set the override, if one is set.
    pub set_by: Option<String>,
    /// When it was set (epoch millis), if one is set.
    pub set_at_millis: Option<f64>,
    /// The selectable tiers with their operator-facing consequences.
    pub tiers: Vec<PolicyTierGql>,
    /// When a change bites.
    pub takes_effect: String,
}

impl From<PolicyDto> for PolicyGql {
    fn from(dto: PolicyDto) -> Self {
        Self {
            mode: dto.mode,
            always_approve: dto.always_approve,
            auto_approve_under_usd: dto.auto_approve_under_usd,
            approval_ttl_hours: dto.approval_ttl_hours as f64,
            manifest_mode: dto.manifest_mode,
            manifest_always_approve: dto.manifest_always_approve,
            manifest_auto_approve_under_usd: dto.manifest_auto_approve_under_usd,
            manifest_approval_ttl_hours: dto.manifest_approval_ttl_hours.map(|hours| hours as f64),
            overridden: dto.overridden,
            set_by: dto.set_by,
            // GraphQL has no 64-bit integer scalar, and epoch millis exceed
            // `Int`. `f64` holds them exactly to 2^53, which is the year 287396
            // — the same widening `usage` already applies to its counters.
            set_at_millis: dto.set_at_millis.map(|ms| ms as f64),
            tiers: dto
                .tiers
                .into_iter()
                .map(|t| PolicyTierGql {
                    value: t.value.to_string(),
                    label: t.label.to_string(),
                    description: t.description.to_string(),
                })
                .collect(),
            takes_effect: dto.takes_effect.to_string(),
        }
    }
}

/// Loads the record and projects it through the REST derivation.
pub(crate) async fn resolve_policy(
    runtime: &Arc<CompanyRuntime>,
) -> async_graphql::Result<PolicyGql> {
    let record = runtime
        .store()
        .load(runtime.id())
        .await?
        .ok_or_else(|| async_graphql::Error::new("company not found"))?;
    Ok(PolicyDto::build(&record).into())
}
